src/solvers/flattening/bv_utils.cpp

```diff
@@ -482,26 +482,45 @@ bvt bv_utilst::shift(const bvt &src, const shiftt s, std::size_t dist)
   bvt result;
   result.resize(src.size());
 
+  // 'dist' is user-controlled, and thus arbitary.
+  // We thus must guard against the case in which i+dist overflows.
+  // We do so by considering the case dist>=src.size().
+
   for(std::size_t i=0; i<src.size(); i++)
   {
     literalt l;
 
     switch(s)
     {
-    case shiftt::LEFT:
+    case shiftt::SHIFT_LEFT:
+      // no underflow on i-dist because of condition dist<=i
       l=(dist<=i?src[i-dist]:const_literal(false));
       break;
 
-    case shiftt::ARIGHT:
-      l=(i+dist<src.size()?src[i+dist]:src[src.size()-1]); // sign bit
+    case shiftt::SHIFT_ARIGHT:
+      // src.size()-i won't underflow as i<src.size()
+      // Then, if dist<src.size()-i, then i+dist<src.size()
+      l=(dist<src.size()-i?src[i+dist]:src[src.size()-1]); // sign bit
+      break;
+
+    case shiftt::SHIFT_LRIGHT:
+      // src.size()-i won't underflow as i<src.size()
+      // Then, if dist<src.size()-i, then i+dist<src.size()
+      l=(dist<src.size()-i?src[i+dist]:const_literal(false));
       break;
 
-    case shiftt::LRIGHT:
-      l=(i+dist<src.size()?src[i+dist]:const_literal(false));
+    case shiftt::ROTATE_LEFT:
+      // prevent overflows by using dist%src.size()
+      l=src[(src.size()+i-dist%src.size())%src.size()];
+      break;
+
+    case shiftt::ROTATE_RIGHT:
+      // prevent overflows by using dist%src.size()
+      l=src[(i+dist%src.size())%src.size()];
```

Comment:
Could you bracket this as `(i+(dist%src.size()))%src.size()` (and likewise above) to avoid the "obvious helpful" refactoring.

Reply:
Added the brackets.

Note that the number of operands is already checked by a DATA_INVARIANT:
inline const shift_exprt &to_shift_expr(const exprt &expr)
{
  DATA_INVARIANT(
    expr.operands().size()==2,
    "Shifts must have two operands");
  return static_cast<const shift_exprt &>(expr);
}